Kubernetes job runner integration test and enhancements

mvdbeek opened the pull request:
- pykube is now a conditional requirement and will be installed if a kubernetes runner is set up
- Submit ajs.job_file instead of tool_script.sh, which allows using standard Galaxy features, like setting env variables per destination, reading stdout/stderr from file, collecting job measurements
- Fixes state changes
- Allows selecting the shell to use for all job runners
- Python 3 fixes for container resolver
- Writes stdout and stderr to standard file locations
- Passes GALAXY_SLOTS and GALAXY_MEMORY_MB and GALAXY_MEMORY_MB_PER_SLOT to wrappers
- Includes integration tests for running jobs, cancelling jobs, executing jobs in mulled containers, detecting exit code, stdout, stderr and passing GALAXY_SLOTS and GALAXY_MEMORY_MB.

mvdbeek committed:
Make pykube a conditional requirement

mvdbeek committed:
Use standard os.environ

mvdbeek committed:
Add basic kubernetes integration test

Currently times out becasue jobs aren't monitored succesfully.

mvdbeek committed:
Propagate selected shell to job_script

mvdbeek committed:
Allow skipping pod_retrial

mvdbeek committed:
Use galaxy's exisiting job_file abstraction

This allows setting environment variables, job instrumentation,
exit code checking and so on.

mvdbeek committed:
Use sh as job shell for kubernetes test, bash is not available in default container

mvdbeek committed:
Drop mutable defaults

mvdbeek committed:
Skip broken string encoding logic

mvdbeek committed:
Remove return statements from queue_job, this is never consumed anywhere

mvdbeek committed:
Remove commented out non-functional code

in kubernetes runner.

mvdbeek committed:
Cleanup of kubernetes exception logging

mvdbeek committed:
Test cancelling jobs, running mulled containers on kubernetes

mvdbeek committed:
Fix test skipping

mvdbeek committed:
Don't write fake stderr file

mvdbeek committed:
Add job properties test

mvdbeek committed:
Capture stdout and stderr for kubernetes runner

There is currently no way to discriminate stdout and stderr
from kubernetes logs (https://github.com/kubernetes/kubernetes/issues/28167),
but Galaxy really needs to know about the stderr.
To work around this  we capture the stdout and stderr into files using
the approach outined in https://stackoverflow.com/a/692407.

mvdbeek committed:
Drop custom fail_job and stderr logging methods

We now log stdout and stderr to files, so this isn't needed.

mvdbeek committed:
Remove user-cancelled jobs from monitor queue

mvdbeek committed:
Add utility class for manipulating byte sizes

mvdbeek committed:
Make kubernetes pass GALAXY_SLOTS and GALAXY_MEMORY_MB

Also drops __transform_cpu_value, incoming values are always
strings, so this wasn't functional in any case.

mvdbeek committed:
Fix typo in config/job_conf.xml.sample_advanced

mvdbeek commented:
Alright, this is ready now. @pcm32 @ilveroluca can you guys give this a try ? I only have a toy kubernetes cluster to play with.

jmchilton commented on lib/galaxy/jobs/command_factory.py:
```diff
@@ -229,24 +233,37 @@ def __init__(self, initial_command=u''):
+mkfifo "$out" "$err"
+trap 'rm "$out" "$err"' EXIT
+tee -a stdout.log < "$out" &
+tee -a stderr.log < "$err" >&2 &""",
```
Not to self - confirmed minimal biocontainers have these utils.

pcm32 replied:
mkfifo and tee are on alpine:3.8, but not trap :-(. We need a replacement for it or handle it somehow. I tried with `docker run -it alpine:3.8 sh` and checked inside, otherwise this would make Galaxy incompatible with tools built on top of alpine.

pcm32 replied:
well I searched by `which` but I think trap could be a shell built-in? Then my comment wouldn't apply of course....

mvdbeek replied:
Are you sure about that? This script works for me on `alpine:3.8`
```
#!/bin/sh

out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
mkfifo "$out" "$err"
trap 'rm "$out" "$err"' EXIT
tee -a stdout.log < "$out" &
tee -a stderr.log < "$err" >&2 &
bwa >"$out" 2>"$err"
echo $? > exit_code
```

also trap seems to be there: 
```
/ # command -v trap
trap
/ #
```

ilveroluca replied:
`trap` should be ok.  It's a shell command compatible with both `sh` and `bash`

```
17:05 [pireddu@slynx ~] docker run -it alpine:3.8 sh
/ # trap 'echo "bye bye"' EXIT
```
Enter Ctrl+d...
```
/ # bye bye
```

nsoranzo replied:
FYI you can use `command -V trap` to find out if it's a builtin, alias or normal executable:
```
$ command -V trap
trap is a shell builtin
```

ilveroluca replied:
Scienziato! :bowing_man: :smile:

jmchilton commented:
It looks good to me, I was going to merge before I saw the above comment. If it makes testing easier let me know and I'll merge it, looks ready to go to me.

mvdbeek commented:
Yeah, next step is a set of re-usable runner tests. I want all runners to make sure they know about Galaxy-isms like GALAXY_SLOTS, stderr etc ... . If bugs arise I'm happy and in a good position to fix them.

pcm32 commented:
I have set up some testing for Galaxy-Kubernetes on travis [here](https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/blob/develop/.travis.yml). We could put variations of this if needed somewhere on the Galaxy tests. It is missing though to have a job execution test (so it is currently more of a test that Galaxy spins up, but should be easy to extend for sending a job and waiting for it). It takes a little while though, in case you have alternatives to travis.

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -252,41 +259,42 @@ def __get_k8s_mountable_volumes(self, job_wrapper):
+            if 'limits' in resources:
+                limits = resources['limits']
+                if 'memory' in limits:
+                    envs.append({'name': 'GALAXY_MEMORY_MB', 'value': str(ByteSize(limits['memory']).to_unit('M', as_string=False))})
```
limits normally come provided as Gi normally, are you sure that this part will work fine? Or ByteSize finds out the unit in which the require/limits are given?

mvdbeek replied:
Right, it can take all the valid Kubernetes strings from https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory and turn them into bytes

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -252,41 +259,42 @@ def __get_k8s_mountable_volumes(self, job_wrapper):
+                if 'memory' in limits:
+                    envs.append({'name': 'GALAXY_MEMORY_MB', 'value': str(ByteSize(limits['memory']).to_unit('M', as_string=False))})
+                if 'cpu' in limits:
+                    envs.append({'name': 'GALAXY_SLOTS', 'value': limits['cpu']})
```
Please note that cpu values can be fractions of CPU, not sure if that is going to play nicely with the usage of $GALAXY_SLOTS in other places. I would take the ceiling of the value of cpu limits to put on GALAXY_SLOTS.

mvdbeek replied:
that's a good point, I'll do that.

mvdbeek replied:
Alright, [I've set the CPU to 1.9 in the job_conf.xml](https://github.com/galaxyproject/galaxy/pull/6958/commits/676bdb5b38ed3032dea836d022ab299014d0df4a) setup and we were already testing that GALAXY_SLOTS is 2.

mvdbeek commented:
I think with this travis setup you should be able to run the integration tests here, which run a couple of jobs that use important Galaxy features. In the best case you'd just have to `sh run_tests.sh -integration test/integration/test_kubernetes_runner.py`

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -336,48 +344,23 @@ def __get_cpu_request(self, job_wrapper):
-           If it is an already formatted string, it is returned as it was.
         """
-        if not isinstance(mem_value, str) and float(mem_value) == 0:
-            return None
```
We don't need to handle this corner case anymore?

mvdbeek replied:
I don't think this could have worked as this was always going to be a string after parsing the destination params.

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -501,63 +484,6 @@ def __job_failed_due_to_low_memory(self, job_state):
-            try:
-                pod = Pod(self._pykube_api, pod_obj)
-                log_string += "\n\n==== Pod " + pod.name + " log start ====\n\n"
-                log_string += pod.logs(timestamps=True)
```
So if we are removing this, where are we now retrieving the stdout of the tools when running on k8s? I couldn't find in the PR any other part where this is happening.

mvdbeek replied:
This is now part of the command line generation -- see `capture_stdout_stderr`. This will intercept the job script output, but continue printing out, so you can still use the kubernetes pod logs if necessary.

mvdbeek committed:
Take ceiling of kubernetes cpu value

pcm32 commented on lib/galaxy/jobs/runners/util/job_script/CLUSTER_SLOTS_STATEMENT.sh:
```diff
@@ -23,6 +23,9 @@ elif [ -f "$PBS_NODEFILE" ]; then
     GALAXY_SLOTS="$LSB_DJOB_NUMPROC"
+elif [ -n "$GALAXY_SLOTS" ]; then
+    # kubernetes runner injects GALAXY_SLOTS into environment
+    $GALAXY_SLOTS=$GALAXY_SLOTS
```
This should be `GALAXY_SLOTS=$GALAXY_SLOTS`, you have an extra $.

mvdbeek replied:
hah, good one. it still worked because we avoid the `else`, I'll fix that in a rebase.

pcm32 reviewed (changes requested):
This is a great improvement, many thanks! I have some questions regarding how certain things are handled now, most notably the logs of the tool (stdout and stderr, which you normally retrieve through kubectl logs). However I found a typo somewhere, so I thought I'd made it a proper review. Unfortunately I made most of the comments outside the review. So my concerns are not blocking of course (but would appreciate some clarification), but the typo at least it is I guess.

mvdbeek commented:
Oh, and if you want to test this using docker for mac you have to set the TEMP variable to something different than `/var/private`, since that can't be mounted by docker. I'm just doing `TMPDIR=~/temp sh run_tests.sh -integration test/integration/test_kubernetes_runner.py`

pcm32 reviewed (approved):
LGTM! Thanks again! I should be able to test this after Thursday if that is fine.

mvdbeek commented:
Sure, I'm not in a hurry for this.

pcm32 commented:
On second thoughts, I think that galaxy slots should be equal to the floor of cpu-limits instead of it's ceiling, otherwise we risk having the job killed by kubernetes for trespassing the max amount of CPUs allocated.

pcm32 commented:
Handling of course the case where this would lead to zero slots, which might produce weird behaviour in the tool.

mvdbeek commented:
Is that how kubernetes handles this ? I'd assume it'd just throttle, otherwise I can't really see how fractions of a cpu would work ?

pcm32 commented:
Yes, you're right, it throttles for CPU, it kills for memory, so yes, no change needed on that!

pcm32 commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -435,44 +418,44 @@ def check_watched_item(self, job_state):
 
             # This assumes jobs dependent on a single pod, single container
             if succeeded > 0:
-                self.__produce_log_file(job_state)
```
Unfortunately in the new setup I fail to see the stdout and stderrr of jobs in Galaxy after execution. Any ideas @mvdbeek? I can confirm that this is viewable before this PR (as I was testing the same tool at both sides.

pcm32 replied:
(I'm not sure the cuprit is actually there, I just pointed to any potentially related piece of the code)

mvdbeek replied:
Are you using a specific tool to check this? I'm a bit surprised by this since the integration test does test visibility of stdout and stderr messages

pcm32 replied:
Well I saw it on a specific tool we are developing to bring data from our archive... but I had just run the same tool and seen its stdout at the older branch, so I know that the tool generates stdout (this is a tool we are developing here: https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/blob/develop/tools/tertiary-analysis/data/retrieve-scxa.xml). I don't think is related to the tool though. I will check with another one very soon.

pcm32 replied:
I see the lack of stdout with other tools as well (haven't been able to see stdout on any tools I have tried).

mvdbeek replied:
I can't reproduce this, this is working fine for me (see the red items, this is the text from stderr):
<img width="218" alt="screen shot 2018-11-23 at 17 01 50" src="https://user-images.githubusercontent.com/6804901/48952418-666b7b00-ef42-11e8-918a-69b664f1d4df.png">

As in other job runner we only produce a message on stdout/stderr if there is actually something logged, the previous implementation was always producing a message with the pod logs I think.
The pod logs for instance shouldn't appear in a tools' stdout/stderr, if necessary the admin can look at those. If the pod fails we get `More pods failed than allowed. See stdout for pods details.`

pcm32 replied:
...interesting... do you manage to see stdout? I couldn't see that either. I did see the Fatal errors like you show when using python tools. So maybe we were just misusing the stderr. The logs were used to produce the stdout, but I agree is much better to be able to separate err and out (logs produced both).

pcm32 reviewed (changes requested):
I have found that stdout and stderr are not visible now from the Galaxy UI on this branch when running tools with k8s.

mvdbeek commented on test/integration/test_kubernetes_runner.py:
```diff
@@ -0,0 +1,229 @@
+        details = self.dataset_populator.get_job_details(result['jobs'][0]['id'], full=True).json()
+
+        assert details['state'] == 'error', details
+        assert details['stdout'] == 'The bool is not true\n', details
```
This test tests the presence of messages on stdout and stderr, so I'd assume this works ?

pcm32 commented:
I also get some strange errors with python tools:
```
python: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

I'm inclined to blame the call of `_galaxy_setup_environment` that now makes it to the script executed inside the container. I don't think that we should meddle in any way with the environment inside the container (ie python env active in this case), besides mere environment variables of course.

Would it be possible to make the script avoid that `_galaxy_setup_environment` or not call it?

mvdbeek commented:
I agree, some things in `_galaxy_setup_environment` aren't necessary in a kubernetes setup, I'll see if we can modularize this. Activating the virtual env would indeed be problematic.

jmchilton commented:
> I'll see if we can modularize this

So this is back in WIP?

mvdbeek commented:
yes

pcm32 committed:
Inject env vars for k8s job to avoid changing tool environment inside container.

pcm32 commented:
Just tried the env vars based fix, but python tools are still throwing an error:

```
Fatal error: Exit code 127 ()
```

with no more details. The container does see the GALAXY_VIRTUAL_ENV and GALAXY_LIB set to None.

pcm32 commented:
...errors are different... maybe setting GALAXY_LIB to None was exaggerated... will try without that..

pcm32 commented:
No, on revision I think is correct to set GALAXY_LIB to None, otherwise we might end up adding virtualenv paths to the containers python path, wrecking python execution (which should depend only on the python setup of the tool container). I'm however confused by:

https://github.com/galaxyproject/galaxy/blob/9b2f3af6f3967ab172cb5bbaa527ee622dd8fcf5/lib/galaxy/jobs/runners/util/job_script/__init__.py#L44

this gives me the impression that already those variables where set to None by default. The other issue is that apparently I should be setting the lowercase version of the variables to None, not the upper case ones... will try that better.

pcm32 commented:
Error remains despite changes to env vars to lowercase.

pcm32 commented:
Non-python tools show no issues... it is with a python tool (that otherwise works on 18.09 without these changes on the same setup).

pcm32 commented:
Ok, the problem seems to be that even though I set env variables at the k8s job level, the variables used to write the script happen much earlier and the main script of the job still gets GALAXY_VIRTUAL_ENV and GALAXY_LIB set to values that are different to None. I went into a failed job and reproduce it but changing manually those variables in the script and that fixed it. So question is, how can the runner control variables GALAXY_VIRTUAL_ENV and GALAXY_LIB (probably the lowercase ones) before they get injected into the main execution script? Setting the job_wrapper.galaxy_lib_dir to None?

pcm32 commented:
Besides setting those variables somehow to None, the other issue is that the function `_galaxy_setup_environment()` gets as well a rogue activation of the galaxy virtualenv which cannot be  turned off in any way:

```
_galaxy_setup_environment() {                                                                                                    
    local _use_framework_galaxy="$1"                                                                                             
    _GALAXY_JOB_HOME_DIR="/export/galaxy-central/database/job_working_directory/000/4/home"                                      
    _GALAXY_JOB_TMP_DIR=$([ ! -e '/export/galaxy-central/database/job_working_directory/000/4/tmp' ] || mv '/export/galaxy-centra
                                                                                                                                 
. "/export/venv/bin/activate"                                                                                                    
    if [ "$GALAXY_LIB" != "None" -a "$_use_framework_galaxy" = "True" ]; then                                                    
...
```

pcm32 commented:
@jmchilton @mvdbeek what would be the correct way forward? I was thinking on adding the following:

- In the job template that generates the bash script for execution I would add some overrides for GALAXY_LIB and GALAXY_VIRTUAL_ENV of the type:

```
GALAXY_LIB=${OVERRIDE_GALAXY_LIB:-"$galaxy_lib"}
GALAXY_VIRTUAL_ENV=${OVERRIDE_GALAXY_VIRTUAL_ENV:-''$galaxy_virtual_env"}
```

so that you can override with either "None" or a value of choice by setting the optional $OVERRIDE_* vars.

- A clause to avoid the `. "/export/venv/bin/activate"` line via a:

```
if [ -z ${IGNORE_GALAXY_VIRTUAL_ENV+x} ]; then
   $env_setup_commands
fi
```  

which would only have an effect if $IGNORE_GALAXY_VIRTUAL_ENV is defined. I would set these variables in particular through the k8s job definition.

Would this be fine @jmchilton @mvdbeek or is there a more elegant approach? I don't see one, and other runners using containers don't seem to be handling this, but they are manually handling the command line generation, as it was the case before for k8s. Please let me know so that we can have this working.

mvdbeek commented:
@pcm32 I'll have a look, there is probably a flag missing for containerization somewhere.

pcm32 commented:
I would suggest as well that the integration tests try to incorporate the case of a python tool being scheduled, to account for this case.

mvdbeek committed:
Import from correct module

mvdbeek committed:
Revert "job script returns tool exit code in any case"

This breaks exit code reporting and leads to pod re-scheduling.

This reverts commit 6beeec836f6c6ab479700e6a1e1accbe5069dacf.

mvdbeek commented:
Actually the single mount point thing makes it difficult to test tools with a wrapper living next to the tool ... :/

mvdbeek committed:
Allow adding multiple volume claims

mvdbeek committed:
Set galaxy_virtual_env to None for kubernetes runner

This prevents attempts of activating Galaxy's virtualenv
inside the container.

mvdbeek committed:
Drop GALAXY_VIRTUAL_ENV env hack

GALAXY_VIRTUAL_ENV is injected through job_wrapper.galaxy_virtual_env,
which does `os.environ.get(VIRTUAL_ENV, None)`, so this didn't work.
The previous commit addresses that.

mvdbeek committed:
Fix typo fixed by reverted commit 6beeec8

mvdbeek commented:
Alright, @pcm32 https://github.com/galaxyproject/galaxy/pull/6958/commits/7e329770ec66cb41518fdebdd873008b33616b4b should fix your issue with GALAXY_VIRTUAL_ENV bleeding into the container. I didn't have any problem also without the commit, but that's because I am only mounting in the test directory and the tool directory, so the virtualenv couldn't be activated. That's fixed now, I hope this works for you as well. Note that I changed the syntax for mounting in PVCs in https://github.com/galaxyproject/galaxy/pull/6958/commits/65f33dfea6fe16d79388e835043f4f33ff6fe363, let me know what you think about that.

mvdbeek committed:
Add test for tool using python dependencies

pcm32 commented:
Thanks, I'll try it as soon as I can. I have some concerns on the multiple mount points (made the comment on that commit), but great that we can have this working!

pcm32 commented:
Unfortunately after the last rebase the images don't manage to spin up galaxy correctly, we must have grabbed a bug in dev. Images build correctly (see docker-galaxy-stable/compose build-orchestration-images.sh), but fail to start with this error:

```
galaxy.jobs DEBUG 2019-02-21 14:36:18,901 [p:1953,w:0,m:0] [MainThread] Done loading job configuration
Traceback (most recent call last):
  File "lib/galaxy/webapps/galaxy/buildapp.py", line 49, in app_factory
    app = galaxy.app.UniverseApplication(global_conf=global_conf, **kwargs)
  File "lib/galaxy/app.py", line 128, in __init__
    self.watchers = ConfigWatchers(self)
  File "lib/galaxy/webapps/galaxy/config_watchers.py", line 34, in __init__
    self.tool_data_watcher = get_tool_data_dir_watcher(self.app.tool_data_tables, config=self.app.config)
  File "lib/galaxy/tools/toolbox/watcher.py", line 57, in get_tool_data_dir_watcher
    observer_class = get_observer_class(config_value, default="False", monitor_what_str="tool-data directory")
  File "lib/galaxy/tools/toolbox/watcher.py", line 46, in get_observer_class
    raise Exception(message)
Exception: Watchdog library unavailable, cannot monitor tool-data directory.
```

as such I cannot yet test this :-( . I will investigate what is the problem next week and try to fix.

Happy to go with the additional mount points, but will probably stay without it for a while in a branch until we move to 19.05, to avoid changing the helm charts (I maintain a number of different instances at different galaxy versions, and I don't have man power to deal with that now).

mvdbeek commented:
This seems to be missing the watchdog library, so you can either run through common_startup.sh or disable watch_tool_data_dir https://github.com/galaxyproject/galaxy/blob/dev/config/galaxy.yml.sample#L402 . I don't think this has changed on the Galaxy side.

pcm32 commented:
Yes, strange, as I haven't changed either the default setting, so it should be turned off (for watchdog). I tried a completely clean start with a fresh PV, but still.

pcm32 commented:
I had to go inside the running container and install watchdog... unfortunately the container fails with the same error:
```
$ kubectl logs galaxy-mk-1809-int-test-2-dwxbm
python: error while loading shared libraries: libz.so.1: cannot open shared object file: No such file or directory
```

if I inspect the script file, generated for the failed job I see:
```
#!/bin/bash



_galaxy_setup_environment() {
    local _use_framework_galaxy="$1"
    _GALAXY_JOB_HOME_DIR="/export/galaxy-central/database/job_working_directory/000/2/home"
    _GALAXY_JOB_TMP_DIR=$([ ! -e '/export/galaxy-central/database/job_working_directory/000/2/tmp' ] || mv '/export/galaxy-central/database/job_working_directory/000/2/tmp' '/export/galaxy-central/database/job_working_directory/000/2'/tmp.$(date +%Y%m%d-%H%M%S) ; mkdir '/export/galaxy-central/database/job_working_directory/000/2/tmp'; echo '/export/galaxy-central/database/job_working_directory/000/2/tmp')
    
. "/export/venv/bin/activate"
    if [ "$GALAXY_LIB" != "None" -a "$_use_framework_galaxy" = "True" ]; then
        if [ -n "$PYTHONPATH" ]; then
            PYTHONPATH="$GALAXY_LIB:$PYTHONPATH"
        else
            PYTHONPATH="$GALAXY_LIB"
        fi
        export PYTHONPATH
    fi
    # These don't get cleaned on a re-run but may in the future.
    [ -z "$_GALAXY_JOB_TMP_DIR" -a ! -f "$_GALAXY_JOB_TMP_DIR" ] || mkdir -p "$_GALAXY_JOB_TMP_DIR"
    [ -z "$_GALAXY_JOB_HOME_DIR" -a ! -f "$_GALAXY_JOB_HOME_DIR" ] || mkdir -p "$_GALAXY_JOB_HOME_DIR"
    if [ "$GALAXY_VIRTUAL_ENV" != "None" -a -f "$GALAXY_VIRTUAL_ENV/bin/activate" \
         -a "`command -v python`" != "$GALAXY_VIRTUAL_ENV/bin/python" ]; then
        . "$GALAXY_VIRTUAL_ENV/bin/activate"
    fi
}


# The following block can be used by the job system
# to ensure this script is runnable before actually attempting
# to run it.
if [ -n "$ABC_TEST_JOB_SCRIPT_INTEGRITY_XYZ" ]; then
    exit 42
fi

export GALAXY_SLOTS_CONFIGURED="1"
if [ -n "$SLURM_CPUS_ON_NODE" ]; then
    # This should be valid on SLURM except in the case that srun is used to
    # submit additional job steps under an existing allocation, which we do not
    # currently do.
    GALAXY_SLOTS="$SLURM_CPUS_ON_NODE"
elif [ -n "$SLURM_NTASKS" ] || [ -n "$SLURM_CPUS_PER_TASK" ]; then
    # $SLURM_CPUS_ON_NODE should be set correctly on SLURM (even on old
    # installations), but keep the $SLURM_NTASKS logic as a backup since this
    # was the previous method under SLURM.
    #
    # Multiply these values since SLURM_NTASKS is total tasks over all nodes.
    # GALAXY_SLOTS maps to CPUS on a single node and shouldn't be used for
    # multi-node requests.
    GALAXY_SLOTS=`expr "${SLURM_NTASKS:-1}" \* "${SLURM_CPUS_PER_TASK:-1}"`
elif [ -n "$NSLOTS" ]; then
    GALAXY_SLOTS="$NSLOTS"
elif [ -n "$NCPUS" ]; then
    GALAXY_SLOTS="$NCPUS"
elif [ -n "$PBS_NCPUS" ]; then
    GALAXY_SLOTS="$PBS_NCPUS"
elif [ -f "$PBS_NODEFILE" ]; then
    GALAXY_SLOTS=`wc -l < $PBS_NODEFILE`
elif [ -n "$LSB_DJOB_NUMPROC" ]; then
    GALAXY_SLOTS="$LSB_DJOB_NUMPROC"
elif [ -n "$GALAXY_SLOTS" ]; then
    # kubernetes runner injects GALAXY_SLOTS into environment
    GALAXY_SLOTS=$GALAXY_SLOTS
else
    GALAXY_SLOTS="1"
    unset GALAXY_SLOTS_CONFIGURED
fi

export GALAXY_SLOTS
GALAXY_VIRTUAL_ENV="None"
_GALAXY_VIRTUAL_ENV="None"
PRESERVE_GALAXY_ENVIRONMENT="False"
GALAXY_LIB="/export/galaxy-central/lib"
_galaxy_setup_environment "$PRESERVE_GALAXY_ENVIRONMENT"
```

so that means that while GALAXY_VIRTUAL_ENV is being correctly set to None, there is still a direct call to `. "/export/venv/bin/activate"` that is not being controlled by any environment variable, messing up the tools python environment. :-(

mvdbeek commented:
I am not seeing this, any chance this is caused by a non-default setting in your job_conf.xml file ?

mvdbeek commented:
e.g https://github.com/galaxyproject/galaxy/blob/dev/config/job_conf.xml.sample_advanced#L605 / https://github.com/galaxyproject/galaxy/blob/dev/config/job_conf.xml.sample_advanced#L606

mvdbeek commented:
So this is what I see:
```
#!/bin/sh



_galaxy_setup_environment() {
    local _use_framework_galaxy="$1"
    _GALAXY_JOB_HOME_DIR="/Users/mvandenb/temp/tmpw_906fen/000/2/home"
    _GALAXY_JOB_TMP_DIR=$([ ! -e '/Users/mvandenb/temp/tmpw_906fen/000/2/tmp' ] || mv '/Users/mvandenb/temp/tmpw_906fen/000/2/tmp' '/Users/mvandenb/temp/tmpw_906fen/000/2'/tmp.$(date +%Y%m%d-%H%M%S) ; mkdir '/Users/mvandenb/temp/tmpw_906fen/000/2/tmp'; echo '/Users/mvandenb/temp/tmpw_906fen/000/2/tmp')
    
SOME_ENV_VAR="42"; export SOME_ENV_VAR
    if [ "$GALAXY_LIB" != "None" -a "$_use_framework_galaxy" = "True" ]; then
        if [ -n "$PYTHONPATH" ]; then
            PYTHONPATH="$GALAXY_LIB:$PYTHONPATH"
        else
            PYTHONPATH="$GALAXY_LIB"
        fi
        export PYTHONPATH
    fi
    # These don't get cleaned on a re-run but may in the future.
    [ -z "$_GALAXY_JOB_TMP_DIR" -a ! -f "$_GALAXY_JOB_TMP_DIR" ] || mkdir -p "$_GALAXY_JOB_TMP_DIR"
    [ -z "$_GALAXY_JOB_HOME_DIR" -a ! -f "$_GALAXY_JOB_HOME_DIR" ] || mkdir -p "$_GALAXY_JOB_HOME_DIR"
    if [ "$GALAXY_VIRTUAL_ENV" != "None" -a -f "$GALAXY_VIRTUAL_ENV/bin/activate" \
         -a "`command -v python`" != "$GALAXY_VIRTUAL_ENV/bin/python" ]; then
        . "$GALAXY_VIRTUAL_ENV/bin/activate"
    fi
}


# The following block can be used by the job system
# to ensure this script is runnable before actually attempting
# to run it.
if [ -n "$ABC_TEST_JOB_SCRIPT_INTEGRITY_XYZ" ]; then
    exit 42
fi

export GALAXY_SLOTS_CONFIGURED="1"
if [ -n "$SLURM_CPUS_ON_NODE" ]; then
    # This should be valid on SLURM except in the case that srun is used to
    # submit additional job steps under an existing allocation, which we do not
    # currently do.
    GALAXY_SLOTS="$SLURM_CPUS_ON_NODE"
elif [ -n "$SLURM_NTASKS" ] || [ -n "$SLURM_CPUS_PER_TASK" ]; then
    # $SLURM_CPUS_ON_NODE should be set correctly on SLURM (even on old
    # installations), but keep the $SLURM_NTASKS logic as a backup since this
    # was the previous method under SLURM.
    #
    # Multiply these values since SLURM_NTASKS is total tasks over all nodes.
    # GALAXY_SLOTS maps to CPUS on a single node and shouldn't be used for
    # multi-node requests.
    GALAXY_SLOTS=`expr "${SLURM_NTASKS:-1}" \* "${SLURM_CPUS_PER_TASK:-1}"`
elif [ -n "$NSLOTS" ]; then
    GALAXY_SLOTS="$NSLOTS"
elif [ -n "$NCPUS" ]; then
    GALAXY_SLOTS="$NCPUS"
elif [ -n "$PBS_NCPUS" ]; then
    GALAXY_SLOTS="$PBS_NCPUS"
elif [ -f "$PBS_NODEFILE" ]; then
    GALAXY_SLOTS=`wc -l < $PBS_NODEFILE`
elif [ -n "$LSB_DJOB_NUMPROC" ]; then
    GALAXY_SLOTS="$LSB_DJOB_NUMPROC"
elif [ -n "$GALAXY_SLOTS" ]; then
    # kubernetes runner injects GALAXY_SLOTS into environment
    GALAXY_SLOTS=$GALAXY_SLOTS
else
    GALAXY_SLOTS="1"
    unset GALAXY_SLOTS_CONFIGURED
fi

export GALAXY_SLOTS
GALAXY_VIRTUAL_ENV="None"
_GALAXY_VIRTUAL_ENV="None"
PRESERVE_GALAXY_ENVIRONMENT="False"
GALAXY_LIB="/Users/mvandenb/src/galaxy/lib"
_galaxy_setup_environment "$PRESERVE_GALAXY_ENVIRONMENT"
GALAXY_PYTHON=`command -v python`
cd /Users/mvandenb/temp/tmpw_906fen/000/2
if [ -n "$SLURM_JOB_ID" ]; then
    GALAXY_MEMORY_MB=`scontrol -do show job "$SLURM_JOB_ID" | sed 's/.*\( \|^\)Mem=\([0-9][0-9]*\)\( \|$\).*/\2/p;d'` 2>memory_statement.log
fi
if [ -n "$SGE_HGR_h_vmem" ]; then
    GALAXY_MEMORY_MB_PER_SLOT=`echo "$SGE_HGR_h_vmem" | sed 's/G$/ * 1024/' | bc | cut -d"." -f1` 2>memory_statement.log
fi

if [ -z "$GALAXY_MEMORY_MB_PER_SLOT" -a -n "$GALAXY_MEMORY_MB" ]; then
    GALAXY_MEMORY_MB_PER_SLOT=$(($GALAXY_MEMORY_MB / $GALAXY_SLOTS))
elif [ -z "$GALAXY_MEMORY_MB" -a -n "$GALAXY_MEMORY_MB_PER_SLOT" ]; then
    GALAXY_MEMORY_MB=$(($GALAXY_MEMORY_MB_PER_SLOT * $GALAXY_SLOTS))
fi
[ "${GALAXY_MEMORY_MB--1}" -gt 0 ] 2>>memory_statement.log && export GALAXY_MEMORY_MB || unset GALAXY_MEMORY_MB
[ "${GALAXY_MEMORY_MB_PER_SLOT--1}" -gt 0 ] 2>>memory_statement.log && export GALAXY_MEMORY_MB_PER_SLOT || unset GALAXY_MEMORY_MB_PER_SLOT

echo "$GALAXY_SLOTS" > '/Users/mvandenb/temp/tmpw_906fen/000/2/__instrument_core_galaxy_slots' 
echo "$GALAXY_MEMORY_MB" > '/Users/mvandenb/temp/tmpw_906fen/000/2/__instrument_core_galaxy_memory_mb' 
date +"%s" > /Users/mvandenb/temp/tmpw_906fen/000/2/__instrument_core_epoch_start
out="${TMPDIR:-/tmp}/out.$$" err="${TMPDIR:-/tmp}/err.$$"
mkfifo "$out" "$err"
trap 'rm "$out" "$err"' EXIT
tee -a stdout.log < "$out" &
tee -a stderr.log < "$err" >&2 & rm -rf working; mkdir -p working; cd working; /Users/mvandenb/temp/tmpw_906fen/000/2/tool_script.sh > '/Users/mvandenb/temp/tmpw_906fen/000/2/galaxy_2.o' 2> '/Users/mvandenb/temp/tmpw_906fen/000/2/galaxy_2.e'; return_code=$?; sh -c "exit $return_code"
echo $? > /Users/mvandenb/temp/tmpw_906fen/000/2/galaxy_2.ec
date +"%s" > /Users/mvandenb/temp/tmpw_906fen/000/2/__instrument_core_epoch_end
```

pcm32 commented:
You rock @mvdbeek! It must be this line:
https://github.com/ebi-gene-expression-group/container-galaxy-sc-tertiary/blob/develop/config/job_conf.xml#L39

I'll rebuild removing that and try again as soon as I can.

pcm32 commented:
Yes, after that it works. Thanks @mvdbeek for all the follow ups and patiently going through all my concerns. This PR is fine for me.

nsoranzo commented on test/functional/tools/samples_tool_conf.xml:
```diff
@@ -60,6 +60,7 @@
   <tool file="job_environment_default.xml" />
   <tool file="job_environment_default_legacy.xml" />
   <tool file="job_environment_explicit_shared_home.xml" />
+  <tool file="galaxy_slots_and_memory.xml" />
```
@mvdbeek Did you forget to add the `galaxy_slots_and_memory.xml` file? Loading of this tool fails and the corresponding `test_slots_and_memory` integration test is skipped.
